src/jenkins.py

```diff
@@ -705,3 +709,195 @@ def get_agent_name(unit_name: str) -> str:
+    """Get plugin name given a plugin info string of format <plugin-shortname> (<plugin-version>).
+
+    Args:
+        plugin_info: Text containing plugin name and plugin version.
+
+    Raises:
+        ValidationError: if the plugin info string does not conform to expected format.
+
+    Returns:
+        The plugin shortname.
+    """
+    match = re.match(PLUGIN_CAPTURE, plugin_info)
+    if not match:
+        raise ValidationError(f"No plugin matched in: {plugin_info}")
+    return match.group(1)
+
+
+def _build_dependencies_lookup(
+    plugin_dependency_outputs: typing.Iterable[str],
+) -> dict[str, tuple[str, ...]]:
+    """Build a lookup table of plugin short name to list of dependency plugin's short names.
+
+    Args:
+        plugin_dependency_outputs: The plugin dependency output from Jenkins Groovy script.
+
+    Returns:
+        The dependency lookup table.
+    """
+    # coverage tool thinks defaultdict() has an exit condition.
+    dependency_lookup: dict[str, tuple[str, ...]] = defaultdict(lambda: ())  # pragma: no cover
+    for line in plugin_dependency_outputs:
+        match = re.match(PLUGIN_LINE_CAPTURE, line)
+        if not match:
+            continue
+        plugin, dependencies = match.group(1), match.group(3)
+        if not dependencies:
+            dependency_lookup[plugin] = ()
+        try:
+            dependency_lookup[plugin] = tuple(
+                _get_plugin_name(dependency) for dependency in dependencies.split(", ")
+            )
+        except ValidationError as exc:
+            logger.error("Invalid plugin dependency, %s", exc)
+            continue
+
+    return dependency_lookup
+
+
+def _set_jenkins_system_message(message: str, container: ops.Container) -> None:
+    """Set a system message on Jenkins.
+
+    Args:
+        message: The system message to display.
+        container: The Jenkins workload container.
+    """
+    jcasc_yaml = container.pull(JCASC_CONFIG_FILE_PATH, encoding="utf-8").read()
+    config = yaml.safe_load(jcasc_yaml)
+    config["jenkins"]["systemMessage"] = message
+    container.push(
+        JCASC_CONFIG_FILE_PATH, yaml.dump(config), encoding="utf-8", user=USER, group=GROUP
+    )
+
+
+def _traverse_dependencies(
+    plugin: str, dependency_lookup: typing.Mapping[str, typing.Iterable[str]], seen: set[str]
+) -> typing.Iterable[str]:
+    """Recursively traverse through plugin dependencies.
+
+    Args:
+        plugin: The plugin to recurse through the dependencies.
+        dependency_lookup: The plugin and its dependencies lookup table.
+        seen: a set indicating whether a plugin has already been visited.
+
+    Yields:
+        Plugin and its dependents.
+    """
+    if plugin in seen or plugin not in dependency_lookup:
+        return
+    yield plugin
+    seen.add(plugin)
+    for dependency in dependency_lookup[plugin]:
+        yield from _traverse_dependencies(dependency, dependency_lookup, seen)
```

Comment:
What about `yield from itertools.chain(_traverse_dependencies(dependency, dependency_lookup, seen) for dependency in dependency_lookup[plugin])`

More info: https://docs.python.org/3/library/itertools.html#itertools.chain.from_iterable

Reply:
This function has been refactored, in my opinion it is better to keep the for-loop in the refactored code to make it readable.